### PR TITLE
feat(mount): request remount when volume mount becomes read only

### PR DESCRIPTION
**Pull request by ChanYiLin**
ref: https://github.com/longhorn/longhorn/issues/6386

~~- Provide a grpc call to check the volume globalMountPath is readonly or not~~
~~- We can extend the function by simply providing more error message for Controller to check~~

- Instance-Manager will periodically monitor the volume mount point and update the engine status
- Engine/Replica will sync the errorMsg from Instance-Manager
- Volume then check the Engine CRD to get the errorMsg

Pros:
using errorMsg to get the information from process allow us to extend the feature for more situations.

### CODE REVIEW

File: pkg/proxy/volume.go [hunk 1]
```diff
@@ -247,3 +253,24 @@ func (p *Proxy) spdkVolumeUnmapMarkSnapChainRemovedSet(ctx context.Context, req
...
 }
+
+func (p *Proxy) VolumeMountStatusCheck(ctx context.Context, req *rpc.VolumeMountStatusRequest) (resp *rpc.VolumeMountStatusResponse, err error) {
+	if req.VolumeName == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "volume name is required")
+	}
+
+	errMsg := ""
+
+	globalMountPath := fmt.Sprintf(types.VolumeGlobalMountPath, sha256.Sum256([]byte(req.VolumeName)))
```
Comment (derekbit) [inline pkg/proxy/volume.go:264]: What does it append with a checksum?

Comment (ChanYiLin) [reply pkg/proxy/volume.go:264]: The global mount path format is as follow
```
/var/lib/kubelet/plugins/kubernetes.io/csi/<csi driver name>/<sha256sum of pv.spec.csi.volumeHandle>/globalmount
```

File: pkg/process/process_manager.go [hunk 2]
```diff
@@ -111,6 +127,60 @@ func (pm *Manager) Shutdown() {
...
+}
+
+func isEngineProcess(p *Process) bool {
+	return p.PortCount == DefaultEnginePortCount
+}
+
+func getVolumeMountPointMap() (map[string]mount.MountPoint, error) {
+	volumeMountPointMap := make(map[string]mount.MountPoint)
+	mounter := mount.New("")
+	mountPoints, _ := mounter.List()
```
Comment (c3y1huang) [inline pkg/process/process_manager.go:169]: Suggest not to swallow the error.

Comment (ChanYiLin) [reply pkg/process/process_manager.go]: Fixed, thanks!

**Comment by ChanYiLin** [comment]
Hi @c3y1huang , I have made some change

I move the periodical condition check to another goroutine
I found if I put the periodical check together with broadcast , they might interfere each other and cause some weird behaviors.

Please take a look, thanks!

File: pkg/process/process_manager.go [hunk 3]
```diff
@@ -111,6 +143,64 @@ func (pm *Manager) Shutdown() {
 	close(pm.shutdownCh)
 }
 
+func (pm *Manager) checkMountPointStatusForEngine() {
+	var processToUpdate []*Process
+	volumeMountPointMap, err := getVolumeMountPointMap()
+	if err != nil {
+		logrus.Errorf("Failed to get all volume mount points")
```
Comment (c3y1huang) [inline pkg/process/process_manager.go:150]: Should include err.

File: pkg/process/process_manager.go [hunk 4]
```diff
@@ -111,6 +143,64 @@ func (pm *Manager) Shutdown() {
...
+		if isEngineProcess(p) && p.State == StateRunning {
+
+			nameSlices := strings.Split(p.Name, "-")
+			volumeNameSHA := sha256.Sum256([]byte(strings.Join(nameSlices[:len(nameSlices)-2], "-")))
+
+			if mp, exists := volumeMountPointMap[hex.EncodeToString(volumeNameSHA[:])]; exists {
+				p.Conditions[types.EngineConditionFilesystemReadOnly] = false
+				if util.IsMountPointReadOnly(mp) {
+					p.Conditions[types.EngineConditionFilesystemReadOnly] = true
+				}
```
Comment (c3y1huang) [inline pkg/process/process_manager.go:165]:
```suggestion
				p.Conditions[types.EngineConditionFilesystemReadOnly] = util.IsMountPointReadOnly(mp)
```

File: pkg/types/types.go [hunk 5]
```diff
@@ -28,3 +28,10 @@ const (
 	InstanceTypeEngine  = "engine"
 	InstanceTypeReplica = "replica"
 )
+
+const (
+	GlobalMountPathPattern = "/host/var/lib/kubelet/plugins/kubernetes.io/csi/driver.longhorn.io/*/globalmount"
+	VolumeGlobalMountPath  = "/host/var/lib/kubelet/plugins/kubernetes.io/csi/driver.longhorn.io/%x/globalmount"
```
Comment (c3y1huang) [inline pkg/types/types.go:34]: Could you point to where `VolumeGlobalMountPath` is used?

**Review by c3y1huang — CHANGES_REQUESTED** [review]
> I move the periodical condition check to another goroutine
I found if I put the periodical check together with broadcast , they might interfere each other and cause some weird behaviors.

Any chance for us to get some insights into the weird behaviors?

File: pkg/types/types.go (hunk 5 above)
Comment (ChanYiLin) [reply pkg/types/types.go]: It is not used in the new implementation, removed, thanks!

**Comment by ChanYiLin** [comment]
Conflict resolved.

File: pkg/process/process_manager.go [hunk 6]
```diff
@@ -104,6 +136,61 @@ func (pm *Manager) startMonitoring() {
...
+	volumeMountPointMap, err := getVolumeMountPointMap()
+	if err != nil {
+		logrus.WithError(err).Warn("Failed to get all volume mount points")
+	}
+
+	pm.lock.RLock()
+	for _, p := range pm.processes {
+		p.lock.Lock()
+		if isEngineProcess(p) && p.State == StateRunning {
+
```
Comment (innobead) [inline pkg/process/process_manager.go:150]: Remove unnecessary empty line.

File: pkg/process/process_manager.go [hunk 7]
```diff
@@ -104,6 +136,61 @@ func (pm *Manager) startMonitoring() {
...
+			nameSlices := strings.Split(p.Name, "-")
+			volumeNameSHA := sha256.Sum256([]byte(strings.Join(nameSlices[:len(nameSlices)-2], "-")))
+
+			if mp, exists := volumeMountPointMap[hex.EncodeToString(volumeNameSHA[:])]; exists {
+				p.Conditions[types.EngineConditionFilesystemReadOnly] = util.IsMountPointReadOnly(mp)
+				processToUpdate = append(processToUpdate, p)
+			}
+		}
+		p.lock.Unlock()
```
Comment (innobead) [inline pkg/process/process_manager.go:159]: Move just next to `p.lock.Lock()`.

File: pkg/process/process_manager.go [hunk 8]
```diff
@@ -104,6 +136,61 @@ func (pm *Manager) startMonitoring() {
...
+	}
+	pm.lock.RUnlock()
```
Comment (innobead) [inline pkg/process/process_manager.go:161]: Move just next to `pm.lock.Lock()`

File: pkg/process/process_manager.go [hunk 9]
```diff
@@ -104,6 +136,61 @@ func (pm *Manager) startMonitoring() {
...
+
+	for _, p := range processToUpdate {
+		p.UpdateCh <- p
+	}
+
+}
+
+func isEngineProcess(p *Process) bool {
+	return p.PortCount == DefaultEnginePortCount
```
Comment (innobead) [inline pkg/process/process_manager.go:170]: This is a weak check. It works now, but I suggest differentiating process type. Let's have a ticket for it later.

File: pkg/util/util.go [hunk 10]
```diff
@@ -135,3 +136,12 @@ func IsSPDKTgtReady(timeout time.Duration) bool {
...
 }
+
+func IsMountPointReadOnly(mp mount.MountPoint) bool {
+	for _, opt := range mp.Opts {
+		if opt == "ro" {
+			return true
+		}
+	}
+	return false
+}
```
Comment (innobead) [inline pkg/util/util.go:147]: We have a very similar stuff in share manager. I would suggest moving the related code to common-libs, so we can reuse that in all places later. Create a ticket for this later.

[/pkg/server/share_manager.go#L240-L254](https://github.com/longhorn/longhorn-share-manager/blob/de8ff1978f5218656a3a76c3cb6d8ea060d52863/pkg/server/share_manager.go#L240-L254)
```go
func (m *ShareManager) hasHealthyVolume() error {
	mountPath := types.GetMountPath(m.volume.Name)
	if err := exec.CommandContext(m.context, "ls", mountPath).Run(); err != nil {
		return fmt.Errorf(UnhealthyErr, mountPath)
	}

	mounter := mount.New("")
	mountPoints, _ := mounter.List()
	for _, mp := range mountPoints {
		if mp.Path == mountPath && isMountPointReadOnly(mp) {
			return fmt.Errorf(ReadOnlyErr, mountPath)
		}
	}
	return nil
}
```

File: pkg/process/process_manager.go [hunk 11]
```diff
@@ -104,6 +136,61 @@ func (pm *Manager) startMonitoring() {
 	}
 }
 
+func (pm *Manager) checkMountPointStatusForEngine() {
```
Comment (innobead) [inline pkg/process/process_manager.go:139]: Suggest adding unit tests, even though it's a private method. This way we can better understand how mount points will be parsed here.

**Review by innobead — CHANGES_REQUESTED** [review]
In general, LGTM. Just a few comments need to resolve.

File: pkg/process/process_manager.go (hunk 9 above)
Comment (ChanYiLin) [reply pkg/process/process_manager.go:170]: We also use it in longhorn-maanger: https://github.com/longhorn/longhorn-manager/blob/master/engineapi/instance_manager.go#L290
but I agree we should have more reliable check

File: pkg/process/process_manager.go (hunk 7 above)
Comment (ChanYiLin) [reply pkg/process/process_manager.go:159]: we iterate every process
lock it and do some check and unlock it
if we put this next to lock
```
p.lock.Lock()
defer p.lock.Unlock()
```

Then every process will be locked in the end and need to wait until function return to unlock all the process
I think that will effect performance and might stuck some operation?

File: pkg/util/util.go (hunk 10 above)
Comment (ChanYiLin) [reply pkg/util/util.go:147]: https://github.com/longhorn/longhorn/issues/7394

File: pkg/process/process_manager.go (hunk 9 above)
Comment (ChanYiLin) [reply pkg/process/process_manager.go:169]: https://github.com/longhorn/longhorn/issues/7393

File: pkg/process/process_manager.go (hunk 11 above)
Comment (ChanYiLin) [reply pkg/process/process_manager.go:138]: I have move the mount point parsing and process name parsing out of the function and make it simpler to understand. With that, I think we don't need unit tests for it now.
Please take a look, thanks!

File: pkg/process/process_manager.go [hunk 12]
```diff
@@ -104,6 +115,64 @@ func (pm *Manager) startMonitoring() {
 	}
 }
 
+func (pm *Manager) startInstanceConditionCheck() {
```
Comment (innobead) [inline pkg/process/process_manager.go:118]: nit: we can merge startInstanceConditionCheck to startMonitor to reduce one goroutine, as it seems not quite necessary to have two here?

File: pkg/process/process_manager.go (hunk 7 above)
Comment (innobead) [reply pkg/process/process_manager.go:167]: yeah, misunderstood the scope of defer stack. No need to change here.

**Review by innobead — APPROVED** [review]
LGTM.

File: pkg/process/process_manager.go (hunk 12 above)
Comment (innobead) [reply pkg/process/process_manager.go:118]: https://github.com/longhorn/longhorn-instance-manager/pull/316#issuecomment-1842403056

**Pull request merged**